tfhe/src/integer/server_key/radix_parallel/tests_unsigned/test_neg.rs

```diff
@@ -0,0 +1,195 @@
+// Unchecked Tests
+//=============================================================================
+
+pub(crate) fn unchecked_neg_test<P, T>(param: P, mut executor: T)
+where
+    P: Into<PBSParameters>,
+    T: for<'a> FunctionExecutor<&'a RadixCiphertext, RadixCiphertext>,
+{
+    let (cks, sks) = KEY_CACHE.get_from_params(param, IntegerKeyKind::Radix);
+    let sks = Arc::new(sks);
+    let cks = RadixClientKey::from((cks, NB_CTXT));
+
+    let mut rng = rand::thread_rng();
+
+    let modulus = unsigned_modulus(cks.parameters().message_modulus(), NB_CTXT as u32);
+
+    let max_noise_level = sks.key.max_noise_level;
+    let max_degree = sks.key.max_degree;
+
+    let trivial0 = sks.create_trivial_radix(0u64, NB_CTXT);
+    executor.setup(&cks, sks);
+
+    // negation involves either scalar operation or pbs, noise is always nominal
+    let expected_noise_levels = ExpectedNoiseLevels::new(NoiseLevel::NOMINAL, NB_CTXT);
+    let mut expected_degrees = ExpectedDegrees::new(Degree::new(0), NB_CTXT);
+
+    for _ in 0..NB_TESTS {
+        let clear = rng.gen::<u64>() % modulus;
+
+        let ctxt = cks.encrypt(clear);
+
+        let encrypted_result = executor.execute(&ctxt);
+
+        expected_noise_levels.panic_if_any_is_not_equal(&encrypted_result);
+        expected_degrees
+            .after_unchecked_neg(&ctxt)
+            .panic_if_any_is_not_equal(&encrypted_result);
+        panic_if_any_block_values_exceeds_its_degree(&encrypted_result, &cks);
+        panic_if_any_block_info_exceeds_max_degree_or_noise(
+            &encrypted_result,
+            max_degree,
+            max_noise_level,
+        );
+
+        let decrypted_result: u64 = cks.decrypt(&encrypted_result);
+        let expected_result = clear.wrapping_neg() % modulus;
+        assert_eq!(decrypted_result, expected_result);
+    }
+
+    // negation of trivial 0
+    {
+        let ct_res = executor.execute(&trivial0);
+        let dec_res: u64 = cks.decrypt(&ct_res);
+        assert_eq!(0, dec_res);
+    }
+}
+
+//=============================================================================
+// Smart Tests
+//=============================================================================
+
+pub(crate) fn smart_neg_test<P, T>(param: P, mut executor: T)
+where
+    P: Into<PBSParameters>,
+    T: for<'a> FunctionExecutor<&'a mut RadixCiphertext, RadixCiphertext>,
+{
+    let (cks, sks) = KEY_CACHE.get_from_params(param, IntegerKeyKind::Radix);
+    let sks = Arc::new(sks);
+    let cks = RadixClientKey::from((cks, NB_CTXT));
+
+    let mut rng = rand::thread_rng();
+
+    let modulus = unsigned_modulus(cks.parameters().message_modulus(), NB_CTXT as u32);
+
+    let max_noise_level = sks.key.max_noise_level;
+    let max_degree = sks.key.max_degree;
+
+    executor.setup(&cks, sks);
+
+    // negation involves either scalar operation or pbs, noise is always nominal
+    let expected_noise_levels = ExpectedNoiseLevels::new(NoiseLevel::NOMINAL, NB_CTXT);
+
```

Comment:
no expected degrees ? Guessin too annoying to manage if input has to be cleaned ?

Reply:
Yes precisely checking the degree/noise is more annoying, maybe we could use the is_neg_possible

Reply:
I let you judge whether we can have more precise checks, we did not have them before anyways so we already have an improvement in the properties we test

Reply:
I think for its the added max noise and max degree checks are ok, there a few other aspect I intentionally left out (like making sure default tests triggers the path where inputs are not clean) and will revisit those after

Reply:
sounds good